File: plot/src/main/java/net/fwitz/math/plot/complex/color/DomainColoringAdvancedPlusLogScale.java
package net.fwitz.math.plot.complex.color;

import net.fwitz.math.complex.Complex;

import static java.lang.Math.floor;
import static java.lang.Math.log;

public class DomainColoringAdvancedPlusLogScale extends DomainColoringAdvanced {
    private static final double LN_2 = log(2);
    private static final double SCALE = 0.25 / LN_2;

    @Override
    protected double filterBrightness(double b, Complex z) {
        double x = z.logabs() * SCALE;
        x = 0.3 + 0.7 * (x - floor(x));
        x *= b;
        return (x < 0.0) ? 0.0 : ((x > 1.0) ? 1.0 : x);
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/binary/complex/color/DomainColoringAdvancedPlusLogScale.kt
package net.fwitz.math.plot.binary.complex.color

import net.fwitz.math.binary.complex.Complex
import kotlin.math.floor
import kotlin.math.ln

object DomainColoringAdvancedPlusLogScale : DomainColoringAdvanced() {
    private val LN_2 = ln(2.0)
    private val SCALE = 0.25 / LN_2

    override fun filterBrightness(b: Double, z: Complex): Double {
        var x: Double = z.logabs * SCALE
        x = 0.3 + 0.7 * (x - floor(x))
        x *= b
        return if (x < 0.0) 0.0 else if (x > 1.0) 1.0 else x
    }
}